### ISSUE TITLE
CS8602/CS8604 odd behaviour flagging in unit tests despite null assertion.

### ISSUE THREAD

**Issue by StevePy** [dotnet/roslyn-analyzers#7476]
For example the following code:

```cs
Assert.That(given.CapturedPaginationViewModel, Is.Not.Null);
Assert.That(given.CapturedPaginationViewModel.CurrentPage, Is.EqualTo(1)); // <- CS8602
Assert.That(given.CapturedPaginationViewModel.PageSize, Is.EqualTo(5));
Assert.That(given.CapturedPaginationViewModel.TotalRecords, Is.EqualTo(1));
```

where "given" is a BDD-style wrapper for the test:

```cs
private class Given : BaseGiven
{
    // ...
    public PaginationViewModel? CapturedPaginationViewModel { get; private set; } = default;
    // ...
}
```

The 2nd line, and only the 2nd line is getting flagged with a CS8602. With a .Net Core 8 build this warning did not appear, it seemed to pick up the `Is.Not.Null` assertion and was satisfied with that.  What is quite unusual is that if I add a null forgiveness "!" to the second line, the warning does *not* migrate to the 3rd or 4th lines, they do not produce the CS8602 warning.

CurrentPage, PageSize, and TotalRecords are all `int` properties of the view model. I don't like the idea of slapping null forgiveness bangs in my unit tests, and I do not like having builds littered with warnings to ignore. Adding bangs to the unit tests is a lesser evil but I wanted to raise this just to determine if this is a unexpected issue that can be fixed, or is there a better way to test scenarios like this. (`Assert.Multiple` had no impact on the warning)

**Comment by elgonzo** [comment]
In all likelihood that is not related to the roslyn-analyzers in this github repro.

Modern testing frameworks often come with their own analyzers. For example, nunit (which looks like the test framework you seem to be using) comes with its own analyzers that would contextually suppress CS8602 if they detect that a prior assert would make sure a reference is not being zero (rule NUnit3001 https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit3001.md)

Keeping with my assumption that you are using nunit, then for some reason this nunit analyzer isn't active on your end. I don't know if it is something with your project setup, an incompatibility of the nunit analyzers with .NET 9, or something else. I'd suggest to check nuints github repositories for any issue reports or release notes that might possibly be related to your problem or the NUnit3001 analyzer rule in general.
